core/arch/arm/kernel/thread.c

```diff
@@ -1250,22 +1252,40 @@ void thread_rpc_free_arg(uint64_t cookie)
 
 void thread_rpc_alloc_arg(size_t size, paddr_t *arg, uint64_t *cookie)
 {
-	paddr_t pa;
+	paddr_t pa, page;
 	uint64_t co;
 	uint32_t rpc_args[THREAD_RPC_NUM_ARGS] = {
 		OPTEE_SMC_RETURN_RPC_ALLOC, size
 	};
+	struct shmem_mapping *mapping = NULL;
 
 	thread_rpc(rpc_args);
 
 	pa = reg_pair_to_64(rpc_args[1], rpc_args[2]);
 	co = reg_pair_to_64(rpc_args[4], rpc_args[5]);
-	if (!check_alloced_shm(pa, size, sizeof(uint64_t))) {
-		thread_rpc_free_arg(co);
-		pa = 0;
-		co = 0;
+
```

Comment:
This would be the use case where a TA is loaded.

For normal world to call back and register an SHM, it would need yet another thread in secure world as we don't support calling back from an RPC on the same thread.